### ISSUE TITLE
UTF-8 characters break diagnostics' underline position

### ISSUE THREAD

**Issue by eshepelyuk** [nvimtools/none-ls.nvim#19]
Because I've recently faced this when working with cspell.nvim
https://github.com/davidmh/cspell.nvim/issues/32

I would like to revive this issue from `null-ls`

https://github.com/jose-elias-alvarez/null-ls.nvim/issues/1630

**Comment by mochaaP** [comment]
Possibly related? <https://github.com/nvimtools/none-ls.nvim/blob/a3f8687352f3574451f4411b48e4de8dfa8bf71a/lua/null-ls/utils/init.lua#L61-L72>

**Comment by eshepelyuk** [comment]
> Possibly related?
> 
> <https://github.com/nvimtools/none-ls.nvim/blob/a3f8687352f3574451f4411b48e4de8dfa8bf71a/lua/null-ls/diagnostics.lua#L45-L90>
> <https://github.com/nvimtools/none-ls.nvim/blob/a3f8687352f3574451f4411b48e4de8dfa8bf71a/lua/null-ls/utils/init.lua#L61-L72>

TBH not sure, i haven't dug that deep into a code.
But the issue still exists in `none-ls`

**Comment by davidmh** [comment]
I've been looking into this issue, I think the problem comes from a conflict between the way Lua interprets strings with multi-byte characters and the way we pass the `col` field through the patterns.

For example, the length for the string: `* example typox` in every other language would be `15`, but Lua counts the bytes in the string, not the number of printable characters. This means that for the same string, lua returns `16` as the length of the string.

The report coming from CSpell also counts only printable characters, so for a file like this:

`test.md`
```markdown
* example typox
· example typox
```

The report will be:

`npx cspell --show-suggestions -c cspell.json lint --language-id markdown test.md`
```
1/1 ./test.md 163.45ms X
./test.md:1:11 - Unknown word (typox) Suggestions: [typo, typos, type, typw, tyro]
./test.md:2:11 - Unknown word (typox) Suggestions: [typo, typos, type, typw, tyro]
```

Both lines have the same column as the start of the unknown word, because CSpell doesn't count bytes when reporting the position of the error.

So when we read the column from the report:

https://github.com/davidmh/cspell.nvim/blob/4a9843fdfc75e26d5518ac750c021850a4ca1098/lua/cspell/diagnostics/parser.lua#L26

We just forward whatever we got from the CSpell report.

The `end_col` ends up with the correct position because we calculate it with the custom `from_quote` adapter, which finds the end column programmatically.

To counter that discrepancy, I plan on using the column reported by CSpell only as an index to start looking for the word reported as an error in the `end_col` function, and mutating the entries table to define the `col` property in the same function.

I have a proof of concept that seems to work as expected, I'll test a few scenarios before I push anything.

IMO, that feels a bit too hacky to keep as a long-term solution, we should look into validating the `col` property in none-ls, maybe here:

https://github.com/nvimtools/none-ls.nvim/blob/fa9be1679db7bf57d44128a018db43da5488a07f/lua/null-ls/helpers/diagnostics.lua#L60-L82

**Comment by davidmh** [comment]
This issue was fixed in #36
